File: PAT/BLevel/1005/1005.go
// 1005 继续(3n+1)猜想
// auguralpha
// 2020-03-25

package main

import (
	"fmt"
	"sort"
)

var N []int
var Mark []bool
var Ans []int

func threeN(m int) {
	for m != 1 {
		if m % 2 == 0 {
			m /= 2
		} else {
			m = (m * 3 + 1) / 2
		}

		Presence(m)
	}
}

func Presence(m int) {
	for i := 0; i < len(N); i ++ {
		if m == N[i] {
			Mark[i] = true
			return
		}
	}
}

func main() {
	var k int
	_, _ = fmt.Scanf("%d", &k)
	for i := 0; i < k; i ++ {
		var m int
		_, _ = fmt.Scanf("%d", &m)
		N = append(N, m)
		Mark = append(Mark, false)
	}

	for i := 0; i < len(N); i ++ {
		threeN(N[i])
	}

	for i := 0; i < len(N); i ++ {
		if !Mark[i] {
			Ans = append(Ans, N[i])
		}
	}
	sort.Slice(Ans, func(i, j int)bool {
		return Ans[i] > Ans[j]
	})
	for i := 0; i < len(Ans); i ++ {
		if i == len(Ans) - 1 {
			fmt.Printf("%d\n", Ans[i])
		} else {
			fmt.Printf("%d ", Ans[i])
		}
	}
}

File: PAT/BLevel/1005/1005.cpp
// 1005 继续(3n+1)猜想
// auguralpha
// 2019-02-11
#include <cstdio>
#include <cstring>
#include <vector>
#include <algorithm>
#include <iostream>
using namespace std;
const int maxn = 110;
bool mark[maxn];
vector<int> ans;
int n[maxn];

void presence(int m, int k);
bool three_n(int m, int k);
bool cmp(int a, int b);

int main() {
    int k;
    scanf("%d", &k);
    for(int i = 0; i < k; ++i) {
        scanf("%d", &n[i]);
    }
    memset(mark, true, sizeof(mark));
    for(int i = 0; i < k; ++i) {
        three_n(n[i], k);
    }

    for(int i = 0; i < k; ++i) {
        if(mark[i]) {
            ans.push_back(n[i]);
        }
    }
    sort(ans.begin(), ans.end(), cmp);
    for(int i = 0; i < ans.size(); ++i) {
        if(i == ans.size() - 1) {
            cout << ans[i] << endl;
        }
        else {
            cout << ans[i] << " ";
        }
    }

    return 0;
}

bool three_n(int m, int k) {
    while(m != 1) {
        if(m % 2 == 0) {
            m /= 2;
        }
        else {
            m = (m * 3 + 1) / 2;
        }

        presence(m, k);
    }
    return true;
}

void presence(int m, int k) {
    for(int i = 0; i < k; ++i) {
        if(m == n[i]) {
            mark[i] = false;
            return;
        }
    }
}

bool cmp(int a, int b) {
    return a > b;
}
